cli/types/cypress-npm-api.d.ts

```diff
@@ -0,0 +1,504 @@
+    /**
+     * Absolute path to the saved image
+     */
+    path: string
+    height: pixels
+    width: pixels
+  }
+
+  /**
+   * Cypress test run result for a single spec.
+  */
+  interface RunResult {
+    /**
+     * Accurate test results collected by Cypress.
+     */
+    stats: {
+      suites: number
+      tests: number
+      passes: number
+      pending: number
+      skipped: number
+      failures: number
+      wallClockStartedAt: dateTimeISO
+      wallClockEndedAt: dateTimeISO
+      wallClockDuration: ms
+    },
+    /**
+     * Reporter name like "spec"
+     */
+    reporter: string
+    /**
+     * This is controlled by the reporter, and Cypress cannot guarantee
+     * the properties. Usually this object has suites, tests, passes, etc
+     */
+    reporterStats: object
+    hooks: HookInformation[]
+    tests: TestResult[]
+    error: string | null
+    video: string | null
+    screenshots: ScreenshotInformation[]
+    /**
+     * information about the spec test file.
+    */
+    spec: {
+      /**
+       * filename like "spec.js"
+       */
+      name: string
+      /**
+       * name relative to the project root, like "cypress/integration/spec.js"
+      */
+      relative: string
+      /**
+       * resolved filename of the spec
+       */
+      absolute: string
+    },
+    shouldUploadVideo: boolean
+  }
+
+  /**
+   * Results returned by the test run.
+   */
+  interface CypressRunResult {
+    startedTestsAt: dateTimeISO
+    endedTestsAt: dateTimeISO
+    totalDuration: ms
+    totalSuites: number
+    totalTests: number
+    totalFailed: number
+    totalPassed: number
+    totalPending: number
+    totalSkipped: number
+    runs: RunResult[]
+    browserPath: string
+    browserName: string
+    browserVersion: string
+    osName: string
+    osVersion: string
+    cypressVersion: string
+    // TODO add resolved object to the configuration
+    config: CypressConfiguration
+  }
+
+  /**
+   * Cypress NPM module interface.
+   * @see https://on.cypress.io/module-api
+   * @example
+    ```
+    const cypress = require('cypress')
+    cypress.run().then(results => ...)
+    ```
+  */
+  interface CypressNpmApi {
+    /**
+     * Execute a headless Cypress test run.
+     * @see https://on.cypress.io/module-api#cypress-run
+     * @example
+     ```
+     const cypress = require('cypress')
+     // runs all spec files matching a wildcard
+     cypress.run({
+       spec: 'cypress/integration/admin*-spec.js'
+     }).then(results => {
+       // inspect results object
+     })
+     ```
+     */
+    run(options?: Partial<CypressRunOptions>): Promise<CypressRunResult>,
+    /**
+     * Opens Cypress GUI. Resolves with void when the
+     * GUI is closed.
+     * @see https://on.cypress.io/module-api#cypress-open
+     */
+    open(options?: Partial<CypressOpenOptions>): Promise<void>
+  }
+
+  // export Cypress NPM module interface
+  const cypress: CypressNpmApi
+  export default cypress
```

Comment:
I think we probably want this to be `export = cypress` rather than `export default` if cypress itsself is a `module.exports = ` rather than an `export default`

Reply:
hmm just tried locally `export = cypress` and getting the following dtslint errors
```
Error: /Users/gleb/git/cypress/cli/types/tests/cypress-npm-api-test.ts:1:8
ERROR: 1:8   expect  TypeScript@next compile error: 
Module '"cypress"' has no default export.
ERROR: 3:1   expect  Expected type to be:
  (options?: Partial<CypressRunOptions> | undefined) => Promise<CypressRunResult>
got:
  any
ERROR: 4:1   expect  Expected type to be:
  (options?: Partial<CypressOpenOptions> | undefined) => Promise<void>
got:
  any
ERROR: 5:22  expect  TypeScript@next compile error: 
Parameter 'results' implicitly has an 'any' type.
ERROR: 6:3   expect  Expected type to be:
  CypressRunResult
got:
  any
ERROR: 8:20  expect  TypeScript@next compile error: 
Parameter 'results' implicitly has an 'any' type.
ERROR: 9:3   expect  Expected type to be:
  CypressRunResult
got:
  any
ERROR: 11:1  expect  Expected type to be:
  Promise<void>
got:
  any
```

Reply:
and this import line showsn an error

<img width="858" alt="Screen Shot 2019-04-03 at 10 09 34 AM" src="https://user-images.githubusercontent.com/2212006/55485534-a29d9f00-55f8-11e9-821b-68311c798f60.png">